core/src/main/java/org/apache/iceberg/DistributedDataBatchScan.java

```diff
@@ -0,0 +1,309 @@
+
+  @Override
+  protected boolean useSnapshotSchema() {
+    return true;
+  }
+
+  @Override
+  protected CloseableIterable<ScanTask> doPlanFiles() {
+    Snapshot snapshot = snapshot();
+
+    List<ManifestFile> dataManifests = findMatchingDataManifests(snapshot);
+    List<ManifestFile> deleteManifests = findMatchingDeleteManifests(snapshot);
+
+    boolean planDataRemotely = shouldPlanRemotely(dataPlanningMode(), dataManifests);
+    boolean planDeletesRemotely = shouldPlanRemotely(deletePlanningMode(), deleteManifests);
+
+    if (planDataRemotely || planDeletesRemotely) {
+      ExecutorService workerPool = newWorkerPool();
+
+      CompletableFuture<CloseableIterable<DataFile>> dataFilesFuture =
+          newDataFilesFuture(dataManifests, planDataRemotely, workerPool);
+
+      CompletableFuture<DeleteFileIndex> deletesFuture =
+          newDeletesFuture(deleteManifests, planDeletesRemotely, workerPool);
+
+      try {
+        CloseableIterable<DataFile> dataFiles = dataFilesFuture.join();
+        DeleteFileIndex deletes = deletesFuture.join();
+        return createTasks(dataFiles, deletes);
+
+      } catch (CompletionException e) {
+        dataFilesFuture.cancel(true /* may interrupt */);
+        deletesFuture.cancel(true /* may interrupt */);
+        throw new RuntimeException("Failed to plan files", e);
+
+      } finally {
+        workerPool.shutdown();
+        releaseResources();
+      }
+
+    } else {
+      return planFilesLocally(dataManifests, deleteManifests);
+    }
+  }
+
+  private CompletableFuture<CloseableIterable<DataFile>> newDataFilesFuture(
+      List<ManifestFile> manifests, boolean planRemotely, ExecutorService workerPool) {
+
+    return CompletableFuture.supplyAsync(
+        () -> {
+          if (planRemotely) {
+            scanMetrics().scannedDataManifests().increment(manifests.size());
+            List<DataFile> dataFiles = planDataRemotely(manifests);
+            int skippedDataFilesCount = totalFilesCount(manifests) - dataFiles.size();
+            scanMetrics().skippedDataFiles().increment(skippedDataFilesCount);
+            return CloseableIterable.withNoopClose(dataFiles);
+          } else {
+            ManifestGroup manifestGroup = newManifestGroup(manifests, ImmutableList.of());
+            return CloseableIterable.transform(manifestGroup.entries(), ManifestEntry::file);
+          }
+        },
+        workerPool);
+  }
+
+  private CompletableFuture<DeleteFileIndex> newDeletesFuture(
+      List<ManifestFile> manifests, boolean planRemotely, ExecutorService workerPool) {
+
+    return CompletableFuture.supplyAsync(
+        () -> {
+          DeleteFileIndex.Builder builder;
+
+          if (planRemotely) {
+            scanMetrics().scannedDeleteManifests().increment(manifests.size());
+            List<DeleteFile> deleteFiles = planDeletesRemotely(manifests);
+            int skippedDeleteFilesCount = totalFilesCount(manifests) - deleteFiles.size();
```

Comment:
`totalFilesCount` includes deleted entries in the manifest, which I think makes this calculation incorrect. A deleted file isn't used because it is deleted, but it wasn't skipped because it does not match.

Reply:
Well, that was my assumption too but then I checked `ManifestReader`.

```
  CloseableIterable<ManifestEntry<F>> liveEntries() {
    return CloseableIterable.filter(
        content == FileType.DATA_FILES
            ? scanMetrics.skippedDataFiles()
            : scanMetrics.skippedDeleteFiles(),
        entries(),
        entry -> entry != null && entry.status() != ManifestEntry.Status.DELETED);
  }
```

Shall we fix both places then?

Reply:
I fixed it here, we should probably fix it in `ManifestReader`.